### ISSUE TITLE
How to handle finalize-deployment during upgrades

### ISSUE THREAD

**Issue by masterzen** [coreos/fedora-coreos-tracker#536]
This is not really a bug, but I would  like to have a bit of help, as I don't know FCOS internals too much.

We are coordinating FCOS upgrades of a kubernetes cluster by using https://github.com/rancher/system-upgrade-controller and a custom script that does this:

```sh
#!/bin/bash
exec 2>&1
set -e -o pipefail

booted_status=$(rpm-ostree status --booted --json | jq '.deployments[] | select(.booted==true)')
current_os_checksum=$(echo "$booted_status" | jq -r '.checksum')
current_os_version=$(echo "$booted_status" | jq -r '.version')

echo ">> checking for upgrade from '$current_os_checksum/$current_os_version' to '$SYSTEM_UPGRADE_PLAN_LATEST_VERSION'"

if [ -n "$SYSTEM_UPGRADE_PLAN_LATEST_VERSION" ] && [ "$SYSTEM_UPGRADE_PLAN_LATEST_VERSION" != "$current_os_version" ]; then
  echo ">> upgrading from $current_os_checksum/$current_os_version to '$SYSTEM_UPGRADE_PLAN_LATEST_VERSION'"
  rpm-ostree deploy --lock-finalization version="$SYSTEM_UPGRADE_PLAN_LATEST_VERSION"

  # get back os checksum from human version:
  next_os_checksum=$(rpm-ostree status --json | jq -r ".deployments[] | select(.version==\"$SYSTEM_UPGRADE_PLAN_LATEST_VERSION\") | .checksum")

  echo ">> finalizing upgrade to $next_os_checksum"
  rpm-ostree finalize-deployment "$next_os_checksum"
  echo ">> upgrade finalized ($?) to $next_os_checksum"
  exit 0
else
  echo ">> no upgrade available - not rebooting"
fi

```

The System Upgrade Controller launches a Job that in turns launches a privileged container per host whose exec is chrooting into the host and running the aforementioned script.

But the problem is that `rpm-ostree finalize-deployment` reboots the machine so quickly that it never returns and the container is killed (at least I never get the `upgrade finalized` log line). Thus k8s notices that the container exited abnormally and restarts it when the host has rebooted. Hopefully it doesn't do anything since the upgrade has already been applied. The main issue is that it takes time to reschedule the container, and during that time the upgrade controller can't start upgrading another host.

Any idea on how I could somehow defer the `finalize-deployment` so that the container exits successfully, but the machine still reboots and finalize the deployment ?

Thanks !

**Comment by lucab** [comment]
You could try with a [delaying shutdown inhibitor lock](https://www.freedesktop.org/software/systemd/man/systemd-inhibit.html), however I don't know exactly how the systemd shutdown API behaves in that cases (i.e. whether the caller gets blocked or not).

However, I believe this is in a intrinsic race and eventually you'll have to pick one of these sides:
 * if your container exits before the finalization, you cannot ensure the reboot actually happened
 * if your container stays around long enough to ensure finalization completes, you cannot reliably signal completion to a remote component (as you are racing with the local teardown of the whole OS)

**Comment by masterzen** [comment]
Thanks @lucab, I understand what are the compromises. 
I'll have a look the shutdown inhibitor lock.

I think I need to study a bit more the System Upgrade Controller to check how it detects that the upgrade was applied and can move to another node. If that's only by the container exiting correctly, I can probably add a signaling system using node labels as the CoreOS Update Operator was doing.

I was also thinking about triggering the reboot from the container, but doing the `finalize` partin a systemd unit activated during shutdown. I need to test.

Thanks!

**Comment by lucab** [comment]
/cc @ibuildthecloud @dweomer

[Zincati lock-based strategy](https://github.com/coreos/zincati/blob/master/docs/usage/updates-strategy.md#lock-based-strategy) approaches this by doing the signaling on the next reboot. This is akin to locksmith/CLUO design, and I have a feeling Rancher controller may also be able to work this way.

Related, there is a bit of inversion of control going on here, which would be nice to clear and share the solution. Strictly speaking there is not need for a privileged pod and chrooting, as Zincati is non-privileged and can be pointed at some kind of k8s-hosted reboot manager. That would also help making sure clients are not incorrectly jumping past [barriers and deadends](https://github.com/coreos/fedora-coreos-tracker/blob/master/metadata/updates/specifications.md#glossary).

**Comment by masterzen** [comment]
Regarding barriers and deadends, Rancher's System Upgrade Controller is agnostic of the versioning system and upgrade method (you can use it to upgrade anything including user land stuff not necessarily involving a machine reboot). 

For the moment, we (operators) control the versioning and when we perform the upgrade. We manually tell SUC to launch an upgrade by modifying the plan specified version to upgrade to (so we care about barriers and such). Then the plan executes the update script to upgrade to the version specified in the plan.

Indeed SUC is doing things differently than Zincati/CLUO. In those systems, the machines themselves are checking if they had to be upgraded. SUC has only a central component that takes the decision to launch an upgrade. This isn't perfect because it supposes that all the machines affected by a plan would have to upgrade to the same version from the same version. This isn't an issue for our own case. There is still the possibility to run many plans in parallel if such heterogeneous fleets. 

SUC also supports polling an url to find if a new version is released, but this isn't compatible with the cincinnati protocol. I started to write a "cincinnati proxy" that on one side is able to ask the Fedora update servers about potential updates, and on the other side obey to the SUC protocol (unfortunately I wasn't able to finish it yet). I eventually would implement this in SUC if the project is interested.

When we moved from CoreOS to FCOS, there wasn't any migration plan for CLUO, which was an issue for us. The problem of Zincati/airlock is that it doesn't yet support Kubernetes. We briefly thought about writing a Fleetlock compatible service that would talk to k8s by reusing parts of CLUO, but we were too short in time, so instead switched to SUC which for us was easier to implement.

Anyway, thanks for your insightful help!

**Comment by lucab** [comment]
> We briefly thought about writing a Fleetlock compatible service that would talk to k8s by reusing parts of CLUO, but we were too short in time, so instead switched to SUC which for us was easier to implement.

I agree SUC is likely a better path forward, but we need to solve the impedance mismatch for Zincati to use it. @masterzen can you please open a ticket to the upstream SUC project to see if we can manage to move this forward, and cc me there?

**Comment by masterzen** [comment]
@lucab, done in  rancher/system-upgrade-controller#87
